File: storm-core/src/clj/backtype/storm/daemon/drpc.clj
;; Licensed to the Apache Software Foundation (ASF) under one
;; or more contributor license agreements.  See the NOTICE file
;; distributed with this work for additional information
;; regarding copyright ownership.  The ASF licenses this file
;; to you under the Apache License, Version 2.0 (the
;; "License"); you may not use this file except in compliance
;; with the License.  You may obtain a copy of the License at
;;
;; http://www.apache.org/licenses/LICENSE-2.0
;;
;; Unless required by applicable law or agreed to in writing, software
;; distributed under the License is distributed on an "AS IS" BASIS,
;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
;; See the License for the specific language governing permissions and
;; limitations under the License.

(ns backtype.storm.daemon.drpc
  (:require [backtype.storm.util :as util]
            [backtype.storm.ui.helpers :as helpers]
            [backtype.storm.config :as config]
            [backtype.storm.daemon.common :as common]
            [backtype.storm.log :refer [log-debug log-warn log-message]]
            [ring.middleware.reload :as reload]
            [compojure.core :refer [routes POST GET]])
  (:import [backtype.storm.security.auth AuthUtils ThriftServer ThriftConnectionType ReqContext]
           [backtype.storm.security.auth.authorizer DRPCAuthorizerBase]
           [backtype.storm.generated DistributedRPC$Iface DistributedRPC$Processor
                                     DRPCRequest DRPCExecutionException DistributedRPCInvocations$Iface
                                     DistributedRPCInvocations$Processor]
           [java.util.concurrent Semaphore ConcurrentLinkedQueue]
           [backtype.storm.daemon Shutdownable]
           [java.net InetAddress]
           [backtype.storm.generated AuthorizationException])
  (:gen-class))

(defn timeout-check-secs [] 5)

(defn acquire-queue [queues-atom function]
  (swap! queues-atom
    (fn [amap]
      (if-not (amap function)
        (assoc amap function (ConcurrentLinkedQueue.))
        amap)))
  (@queues-atom function))

(defn check-authorization
  ([aclHandler mapping operation context]
    (if aclHandler
      (let [context (or context (ReqContext/context))]
        (if-not (.permit aclHandler context operation mapping)
          (let [principal (.principal context)
                user (if principal (.getName principal) "unknown")]
              (throw (AuthorizationException.
                       (str "DRPC request '" operation "' for '"
                            user "' user is not authorized"))))))))
  ([aclHandler mapping operation]
    (check-authorization aclHandler mapping operation (ReqContext/context))))

;; TODO: change this to use TimeCacheMap
(defn service-handler [conf]
  (let [drpc-acl-handler (common/mk-authorization-handler (conf config/DRPC-AUTHORIZER) conf)
        ctr (atom 0)
        id->sem (atom {})
        id->result (atom {})
        id->start (atom {})
        id->function (atom {})
        id->request (atom {})
        request-queues (atom {})
        cleanup (fn [id] (swap! id->sem dissoc id)
                  (swap! id->result dissoc id)
                  (swap! id->function dissoc id)
                  (swap! id->request dissoc id)
                  (swap! id->start dissoc id))
        my-ip (.getHostAddress (InetAddress/getLocalHost))  ;; TODO: is this needed?
        clear-thread (util/async-loop
                       (fn []
                         (doseq [[id start] @id->start]
                           (when (> (util/time-delta start) (conf config/DRPC-REQUEST-TIMEOUT-SECS))
                             (when-let [sem (@id->sem id)]
                               (.remove (acquire-queue request-queues (@id->function id)) (@id->request id))
                               (log-warn "Timeout DRPC request id: " id " start at " start)
                               (.release sem))
                             (cleanup id)))
                         (timeout-check-secs)))]
    (reify DistributedRPC$Iface
      (^String execute
        [this ^String function ^String args]
        (log-debug "Received DRPC request for " function " (" args ") at " (System/currentTimeMillis))
        (check-authorization drpc-acl-handler
                             {DRPCAuthorizerBase/FUNCTION_NAME function}
                             "execute")
        (let [id (str (swap! ctr (fn [v] (mod (inc v) 1000000000))))
              ^Semaphore sem (Semaphore. 0)
              req (DRPCRequest. args id)
              ^ConcurrentLinkedQueue queue (acquire-queue request-queues function)]
          (swap! id->start assoc id (util/current-time-secs))
          (swap! id->sem assoc id sem)
          (swap! id->function assoc id function)
          (swap! id->request assoc id req)
          (.add queue req)
          (log-debug "Waiting for DRPC result for " function " " args " at " (System/currentTimeMillis))
          (.acquire sem)
          (log-debug "Acquired DRPC result for " function " " args " at " (System/currentTimeMillis))
          (let [result (@id->result id)]
            (cleanup id)
            (log-debug "Returning DRPC result for " function " " args " at " (System/currentTimeMillis))
            (if (instance? DRPCExecutionException result)
              (throw result)
              (if (nil? result)
                (throw (DRPCExecutionException. "Request timed out"))
                result)))))

      DistributedRPCInvocations$Iface

      (^void result
        [this ^String id ^String result]
        (when-let [func (@id->function id)]
          (check-authorization drpc-acl-handler
                               {DRPCAuthorizerBase/FUNCTION_NAME func}
                               "result")
          (let [^Semaphore sem (@id->sem id)]
            (log-debug "Received result " result " for " id " at " (System/currentTimeMillis))
            (when sem
              (swap! id->result assoc id result)
              (.release sem)
              ))))

      (^void failRequest
        [this ^String id]
        (when-let [func (@id->function id)]
          (check-authorization drpc-acl-handler
                               {DRPCAuthorizerBase/FUNCTION_NAME func}
                               "failRequest")
          (let [^Semaphore sem (@id->sem id)]
            (when sem
              (swap! id->result assoc id (DRPCExecutionException. "Request failed"))
              (.release sem)))))

      (^DRPCRequest fetchRequest
        [this ^String func]
        (check-authorization drpc-acl-handler
                             {DRPCAuthorizerBase/FUNCTION_NAME func}
                             "fetchRequest")
        (let [^ConcurrentLinkedQueue queue (acquire-queue request-queues func)
              ret (.poll queue)]
          (if ret
            (do (log-debug "Fetched request for " func " at " (System/currentTimeMillis))
              ret)
            (DRPCRequest. "" ""))))

      Shutdownable

      (shutdown
        [this]
        (.interrupt clear-thread)))))

(defn handle-request [handler]
  (fn [request]
    (handler request)))

(defn webapp [handler http-creds-handler]
  (->
    (routes
      (POST "/drpc/:func" [:as {:keys [body servlet-request]} func & m]
        (let [args (slurp body)]
          (if http-creds-handler
            (.populateContext http-creds-handler (ReqContext/context)
                              servlet-request))
          (.execute handler func args)))
      (POST "/drpc/:func/" [:as {:keys [body servlet-request]} func & m]
        (let [args (slurp body)]
          (if http-creds-handler
            (.populateContext http-creds-handler (ReqContext/context)
                              servlet-request))
          (.execute handler func args)))
      (GET "/drpc/:func/:args" [:as {:keys [servlet-request]} func args & m]
          (if http-creds-handler
            (.populateContext http-creds-handler (ReqContext/context)
                              servlet-request))
          (.execute handler func args))
      (GET "/drpc/:func/" [:as {:keys [servlet-request]} func & m]
          (if http-creds-handler
            (.populateContext http-creds-handler (ReqContext/context)
                              servlet-request))
          (.execute handler func ""))
      (GET "/drpc/:func" [:as {:keys [servlet-request]} func & m]
          (if http-creds-handler
            (.populateContext http-creds-handler (ReqContext/context)
                              servlet-request))
          (.execute handler func "")))
    (reload/wrap-reload '[backtype.storm.daemon.drpc])
    handle-request))

(defn launch-server!
  ([]
    (let [conf (config/read-storm-config)
          worker-threads (int (conf config/DRPC-WORKER-THREADS))
          queue-size (int (conf config/DRPC-QUEUE-SIZE))
          drpc-http-port (int (conf config/DRPC-HTTP-PORT))
          drpc-port (int (conf config/DRPC-PORT))
          drpc-service-handler (service-handler conf)
          ;; requests and returns need to be on separate thread pools, since calls to
          ;; "execute" don't unblock until other thrift methods are called. So if
          ;; 64 threads are calling execute, the server won't accept the result
          ;; invocations that will unblock those threads
          handler-server (when (> drpc-port 0)
                           (ThriftServer. conf
                             (DistributedRPC$Processor. drpc-service-handler)
                             ThriftConnectionType/DRPC))
          invoke-server (ThriftServer. conf
                          (DistributedRPCInvocations$Processor. drpc-service-handler)
                          ThriftConnectionType/DRPC_INVOCATIONS)
          http-creds-handler (AuthUtils/GetDrpcHttpCredentialsPlugin conf)]
      (util/add-shutdown-hook-with-force-kill-in-1-sec (fn []
                                                    (if handler-server (.stop handler-server))
                                                    (.stop invoke-server)))
      (log-message "Starting Distributed RPC servers...")
      (future (.serve invoke-server))
      (when (> drpc-http-port 0)
        (let [app (webapp drpc-service-handler http-creds-handler)
              filter-class (conf config/DRPC-HTTP-FILTER)
              filter-params (conf config/DRPC-HTTP-FILTER-PARAMS)
              filters-confs [{:filter-class filter-class
                              :filter-params filter-params}]
              https-port (int (conf config/DRPC-HTTPS-PORT))
              https-ks-path (conf config/DRPC-HTTPS-KEYSTORE-PATH)
              https-ks-password (conf config/DRPC-HTTPS-KEYSTORE-PASSWORD)
              https-ks-type (conf config/DRPC-HTTPS-KEYSTORE-TYPE)
              https-key-password (conf config/DRPC-HTTPS-KEY-PASSWORD)
              https-ts-path (conf config/DRPC-HTTPS-TRUSTSTORE-PATH)
              https-ts-password (conf config/DRPC-HTTPS-TRUSTSTORE-PASSWORD)
              https-ts-type (conf config/DRPC-HTTPS-TRUSTSTORE-TYPE)
              https-want-client-auth (conf config/DRPC-HTTPS-WANT-CLIENT-AUTH)
              https-need-client-auth (conf config/DRPC-HTTPS-NEED-CLIENT-AUTH)]

          (helpers/storm-run-jetty
           {:port drpc-http-port
            :configurator (fn [server]
                            (helpers/config-ssl server
                                        https-port
                                        https-ks-path
                                        https-ks-password
                                        https-ks-type
                                        https-key-password
                                        https-ts-path
                                        https-ts-password
                                        https-ts-type
                                        https-need-client-auth
                                        https-want-client-auth)
                            (helpers/config-filter server app filters-confs))})))
      (when handler-server
        (.serve handler-server)))))

(defn -main []
  (util/setup-default-uncaught-exception-handler)
  (launch-server!))

File: storm-core/src/jvm/org/apache/storm/daemon/drpc/DRPC.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.storm.daemon.drpc;

import java.security.Principal;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.atomic.AtomicLong;

import org.apache.storm.Config;
import org.apache.storm.daemon.StormCommon;
import org.apache.storm.generated.AuthorizationException;
import org.apache.storm.generated.DRPCExceptionType;
import org.apache.storm.generated.DRPCExecutionException;
import org.apache.storm.generated.DRPCRequest;
import org.apache.storm.logging.ThriftAccessLogger;
import org.apache.storm.metric.StormMetricsRegistry;
import org.apache.storm.security.auth.IAuthorizer;
import org.apache.storm.security.auth.ReqContext;
import org.apache.storm.security.auth.authorizer.DRPCAuthorizerBase;
import org.apache.storm.utils.Utils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.codahale.metrics.Meter;
import com.google.common.annotations.VisibleForTesting;

public class DRPC implements AutoCloseable {
    private static final Logger LOG = LoggerFactory.getLogger(DRPC.class);
    private static final DRPCRequest NOTHING_REQUEST = new DRPCRequest("","");
    private static final DRPCExecutionException TIMED_OUT = new DRPCExecutionException("Timed Out");
    private static final DRPCExecutionException SHUT_DOWN = new DRPCExecutionException("Server Shutting Down");
    private static final DRPCExecutionException DEFAULT_FAILED = new DRPCExecutionException("Request failed");
    static {
        TIMED_OUT.set_type(DRPCExceptionType.SERVER_TIMEOUT);
        SHUT_DOWN.set_type(DRPCExceptionType.SERVER_SHUTDOWN);
        DEFAULT_FAILED.set_type(DRPCExceptionType.FAILED_REQUEST);
    }
    private static final Meter meterServerTimedOut = StormMetricsRegistry.registerMeter("drpc:num-server-timedout-requests");
    private static final Meter meterExecuteCalls = StormMetricsRegistry.registerMeter("drpc:num-execute-calls");
    private static final Meter meterResultCalls = StormMetricsRegistry.registerMeter("drpc:num-result-calls");
    private static final Meter meterFailRequestCalls = StormMetricsRegistry.registerMeter("drpc:num-failRequest-calls");
    private static final Meter meterFetchRequestCalls = StormMetricsRegistry.registerMeter("drpc:num-fetchRequest-calls");
    
    private static IAuthorizer mkAuthorizationHandler(String klassname, Map<String, Object> conf) {
        try {
            return StormCommon.mkAuthorizationHandler(klassname, conf);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
    
    @VisibleForTesting
    static void checkAuthorization(ReqContext reqContext, IAuthorizer auth, String operation, String function) throws AuthorizationException {
        if (reqContext != null) {
            ThriftAccessLogger.logAccessFunction(reqContext.requestID(), reqContext.remoteAddress(), reqContext.principal(), operation, function);
        }
        if (auth != null) {
            Map<String, String> map = new HashMap<>();
            map.put(DRPCAuthorizerBase.FUNCTION_NAME, function);
            if (!auth.permit(reqContext, operation, map)) {
                Principal principal = reqContext.principal();
                String user = (principal != null) ? principal.getName() : "unknown";
                throw new AuthorizationException("DRPC request '" + operation + "' for '" + user + "' user is not authorized");
            }
        }
    }
    
    //Waiting to be fetched
    private final ConcurrentHashMap<String, ConcurrentLinkedQueue<OutstandingRequest>> _queues = 
            new ConcurrentHashMap<>();
    //Waiting to be returned
    private final ConcurrentHashMap<String, OutstandingRequest> _requests = 
            new ConcurrentHashMap<>();
    private final Timer _timer = new Timer();
    private final AtomicLong _ctr = new AtomicLong(0);
    private final IAuthorizer _auth;
    
    public DRPC(Map<String, Object> conf) {
        this(mkAuthorizationHandler((String)conf.get(Config.DRPC_AUTHORIZER), conf),
                Utils.getInt(conf.get(Config.DRPC_REQUEST_TIMEOUT_SECS), 600) * 1000);
    }
    
    public DRPC(IAuthorizer auth, long timeoutMs) {
        _auth = auth;
        _timer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                cleanupAll(timeoutMs, TIMED_OUT);
            }
        }, timeoutMs/2, timeoutMs/2);
    }
    
    
    private void checkAuthorization(String operation, String function) throws AuthorizationException {
        checkAuthorization(ReqContext.context(), _auth, operation, function);
    }
    
    private void cleanup(String id) {
        OutstandingRequest req = _requests.remove(id);
        if (req != null && !req.wasFetched()) {
            _queues.get(req.getFunction()).remove(req);
        }
    }

    private void cleanupAll(long timeoutMs, DRPCExecutionException exp) {
        for (Entry<String, OutstandingRequest> e : _requests.entrySet()) {
            OutstandingRequest req = e.getValue();
            if (req.isTimedOut(timeoutMs)) {
                req.fail(exp);
                cleanup(e.getKey());
                meterServerTimedOut.mark();
            }
        }
    }
    
    private String nextId() {
        return String.valueOf(_ctr.incrementAndGet());
    }

    private ConcurrentLinkedQueue<OutstandingRequest> getQueue(String function) {
        ConcurrentLinkedQueue<OutstandingRequest> queue = _queues.get(function);
        if (queue == null) {
            _queues.putIfAbsent(function, new ConcurrentLinkedQueue<>());
            queue = _queues.get(function);
        }
        return queue;
    }

    public void returnResult(String id, String result) throws AuthorizationException {
        meterResultCalls.mark();
        LOG.debug("Got a result {} {}", id, result);
        OutstandingRequest req = _requests.get(id);
        if (req != null) {
            checkAuthorization("result", req.getFunction());
            req.returnResult(result);
        }
    }

    public DRPCRequest fetchRequest(String functionName) throws AuthorizationException {
        meterFetchRequestCalls.mark();
        checkAuthorization("fetchRequest", functionName);
        ConcurrentLinkedQueue<OutstandingRequest> q = getQueue(functionName);
        OutstandingRequest req = q.poll();
        if (req != null) {
            req.fetched();
            DRPCRequest ret = req.getRequest();
            return ret;
        }
        return NOTHING_REQUEST;
    }

    public void failRequest(String id, DRPCExecutionException e) throws AuthorizationException {
        meterFailRequestCalls.mark();
        LOG.debug("Got a fail {}", id);
        OutstandingRequest req = _requests.get(id);
        if (req != null) {
            checkAuthorization("failRequest", req.getFunction());
            if (e == null) {
                e = DEFAULT_FAILED;
            }
            req.fail(e);
        }
    }

    public <T extends OutstandingRequest> T execute(String functionName, String funcArgs, RequestFactory<T> factory) throws AuthorizationException {
        meterExecuteCalls.mark();
        checkAuthorization("execute", functionName);
        String id = nextId();
        LOG.debug("Execute {} {}", functionName, funcArgs);
        T req = factory.mkRequest(functionName, new DRPCRequest(funcArgs, id));
        _requests.put(id, req);
        ConcurrentLinkedQueue<OutstandingRequest> q = getQueue(functionName);
        q.add(req);
        return req;
    }
    
    public String executeBlocking(String functionName, String funcArgs) throws DRPCExecutionException, AuthorizationException {
        BlockingOutstandingRequest req = execute(functionName, funcArgs, BlockingOutstandingRequest.FACTORY);
        try {
            LOG.debug("Waiting for result {} {}",functionName, funcArgs);
            return req.getResult();
        } catch (DRPCExecutionException e) {
            throw e;
        } finally {
            cleanup(req.getRequest().get_request_id());
        }
    }

    @Override
    public void close() {
        _timer.cancel();
        cleanupAll(0, SHUT_DOWN);
    }
}
